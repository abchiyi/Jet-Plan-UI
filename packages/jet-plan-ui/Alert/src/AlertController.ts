import type { alert as Alert } from "./Alert";
import { reactive, watch } from "vue";
export default class {
  alerts: Alert[];
  counter: number;
  key: number;

  constructor() {
    this.alerts = reactive([]);
    this.counter = 0;
    this.key = 0;

    watch(this.alerts, alerts => {
      const alert = alerts[0] as Alert | undefined;
      this.key = alert ? alert.key : 0;
      if (alert && alert.timeout) {
        setTimeout(() => this._$remove(alert), alert.timeout);
      }
    });
  }

  sendMessage(
    messageType: Alert["type"],
    timeout: Alert["timeout"],
    content: Alert["content"]
  ) {
    this.counter++;
    const data = {
      type: messageType,
      key: this.counter,
      content,
      timeout,
    };
    this.alerts.push(data);
  }

  _$remove(data: Alert) {
    const index = this.alerts.indexOf(data);
    if (index !== -1) this.alerts.splice(index, 1);
  }

  async removeAll(ms?: number) {
    function sleep(delay: number) {
      return new Promise(resolve => setTimeout(resolve, delay));
    }
    if (ms) {
      while (this.alerts.length) {
        const data = this.alerts.shift();
        if (!data) break;
        this._$remove(data);
        await sleep(ms);
      }
    } else {
      this.alerts.splice(0);
    }
    this.counter = 0; // 完全清空序列后重置计数器
  }
}
