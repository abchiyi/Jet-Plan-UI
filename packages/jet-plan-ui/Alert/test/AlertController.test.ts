import AlertController from "../src/AlertController";
import { describe, test, expect } from "vitest";
import { h } from "vue";

function sleep(delay: number) {
  return new Promise(resolve => setTimeout(resolve, delay));
}
const alertController = new AlertController();

describe("Test AlertController ", () => {
  test("拥有属性 alerts ", () => {
    expect(alertController).toHaveProperty("alerts");
  });

  test("Method: sendMessage", async () => {
    await alertController.removeAll();

    alertController.sendMessage("success", 50, "Test1");
    alertController.sendMessage("success", 50, "Test2");

    expect(alertController.alerts[0]).toMatchSnapshot();

    expect(alertController.alerts.length).toBe(2);

    await sleep(60);
    expect(alertController.alerts.length).toBe(1);
    await sleep(60);
    expect(alertController.alerts.length).toBe(0);

 
  });

  test("Method: _$remove", async () => {
    await alertController.removeAll();
    alertController.sendMessage("success", null, "message2");
    alertController._$remove(alertController.alerts[0]);
    expect(alertController.alerts.length).toBe(0);
    expect(alertController.alerts).not.include({
      type: "success",
      content: "Test",
      timeout: null,
    });
  });

  test("Method: removeAll", async () => {
    await alertController.removeAll();
    expect(alertController.alerts.length).toBe(0);

    for (let n = 0; n < 15; n++) {
      alertController.sendMessage("success", 1000, "Test");
    }
    await alertController.removeAll(200);
    expect(alertController.alerts.length).toBe(0);
    expect(alertController.counter).toEqual(0);
  });

  test("Attr:key", async () => {
    await alertController.removeAll();

    alertController.sendMessage("info", null, "test1");
    await sleep(2);
    expect(alertController.key).toBe(1);
    expect(alertController.key).toBe(alertController.alerts[0].key);

    // key 的值始终与 .alerts 中的第一位元素保持一致
    alertController.sendMessage("info", null, "test2");
    alertController.alerts.shift();
    await sleep(2);
    expect(alertController.key).toEqual(2);
  });
  test("Attr:counter", async () => {
    await alertController.removeAll();
    alertController.counter = 0;

    alertController.sendMessage("info", null, "test1");
    expect(alertController.counter).toEqual(1);
  });
});
