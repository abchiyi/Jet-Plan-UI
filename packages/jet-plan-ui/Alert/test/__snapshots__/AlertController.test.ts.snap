// Vitest Snapshot v1

exports[`Test AlertController  > Method: sendMessage 1`] = `
{
  "content": "Test1",
  "key": 1,
  "timeout": 50,
  "type": "success",
}
`;
